sherpa_contrib/diag_resp.py

```diff
@@ -0,0 +1,908 @@
+
+        if telescope == "chandra" and instrument == "ACIS" and self.detnam is None:
+            return chantype
+
+        return None
+
+
+    def _get_ebounds_blk(self, instrument: str|None = None,
+                         detnam: str|None = None,
+                         instfilter: str|None = None,
+                         subchan: int|None = None,
+                         chantype: str|None = None) -> str:
+
+        detstr = f"{detnam:->{len(detnam)+1}}" if detnam is not None else ""
+        filtstr = f"{instfilter:/>{len(str(instfilter))+1}}" if instfilter is not None else ""
+        nchanstr = f"{subchan:_>{len(str(subchan))+1}}chan" if subchan is not None else ""
+        chantypestr = f"_{chantype}" if chantype is not None else ""
+
+        blkname = f"{instrument}{detstr}{filtstr}{nchanstr}{chantypestr}"
+
+        return blkname
+
+
+    def get_acis_egrid(self, chantype: str = "PI") -> tuple[npt.NDArray, npt.NDArray, int]:
+        """
+        return energy grid and channel offset for Chandra/ACIS detector
+        """
+
+        offset: int = 1
+
+        chan = chantype.upper()
+
+        if chan not in ["PI","PHA","PHA_NO-CTICORR"]:
+            logger = getLogger(__name__)
+            logger.warning("Warning: An invalid 'chantype' provided!  Assuming ACIS PI spectral channel type...")
+
+        if chan.startswith("PHA"):
+            detchans: int = 4096
+
+            if chan == "PHA":
+                ebin: float = 4.460 # eV CTI corrected
+            else:
+                ebin: float = 4.485 # eV non-CTI corrected
+        else:
+            detchans: int = 1024
+            ebin: float = 14.6 # eV
+
+        emin = arange(detchans) * (ebinkeV := ebin/1000) # in keV
+        emax = emin + ebinkeV # in keV
+
+        return emin, emax, offset
+
+
+    def get_egrid(self, telescope: str = "",
+                  instrument: str|None = None,
+                  detnam: str|None = None,
+                  instfilter: str|None = None,
+                  subchan: int|None = None,
+                  chantype: str|None = None) -> tuple[npt.NDArray, npt.NDArray, int]:
+        """
+        return energy grid and channel offset for a given telescope
+        and instrument configuration
+        """
+
+        from sherpa.astro.io import backend as shp_backend
+
+        fn = f"{os.environ['ASCDS_INSTALL']}/data/ebounds-lut/{self.telescope}-ebounds-lut.fits"
+
+        chantype = self._set_chantype_none(telescope = self.telescope,
+                                           instrument = self.instrument,
+                                           chantype = self.chantype)
+
+        blkname = self._get_ebounds_blk(instrument = instrument,
+                                        detnam = detnam,
+                                        instfilter = instfilter,
+                                        subchan = subchan,
+                                        chantype = chantype)
+
+        try:
```

Comment:
So, we should be able to just use the sherpa io code to read this in (in a round-about manner). The following has not been tested

```
from sherpa.astro.io import read_table_blocks
```

For example

```
sherpa-4.16.0> from sherpa.astro.io import read_table_blocks

sherpa-4.16.0> infile = "data/ebounds-lut/xrism-ebounds-lut.fits"

sherpa-4.16.0> _, blocks, hdrs = read_table_blocks(infile)
```

Now, blocks and hdrs are dictionaries with keys 1, ..., nblocks - e.g.

```
sherpa-4.16.0> blocks.keys()
dict_keys([1, 2, 3, 4, 5])
```

Unfortunately it's not easy to find the block, since crates eats up the HDUNAME/EXTNAME values, so (for Crates) we have

```
sherpa-4.16.0> from sherpa.astro import io

sherpa-4.16.0> io.backend.__name__
'sherpa.astro.io.crates_backend'

sherpa-4.16.0> hdrs[2]
{'TELESCOP': 'HITOMI',
 'INSTRUME': 'SXI',
 'DETNAM': 'PIXEL',
 'FILTER': '',
 'CHANTYPE': 'PI',
 'HDUCLASS': 'OGIP',
 'HDUCLAS1': 'RESPONSE',
 'HDUCLAS2': 'EBOUNDS',
 'HDUVERS': '1.2.0',
 'HDUVERS1': '1.1.0',
 'HDUVERS2': '1.2.0',
 'RMFVERSN': '1992a',
 'CCLS0001': 'CPF',
 'CCNM0001': 'EBOUNDS',
 'CDTP0001': 'DATA',
 'CVSD0001': '2016-03-20',
 'CVST0001': '14:40:33',
 'CDES0001': 'Energy boundaries of spectral channels',
 'CISTATUS': '1',
 'DETCHANS': '4096',
 'CHECKSUM': 'Z5AYb42YZ49Yb49Y',
 'DATASUM': '3744066889',
 'EUNIT': 'keV'}
```

whereas if we used AstroPy we would get

```
sherpa-4.16.0> import sherpa.astro.io.pyfits_backend

sherpa-4.16.0> io.backend = sherpa.astro.io.pyfits_backend

sherpa-4.16.0> _, blocks2, hdrs2 = read_table_blocks(infile)

sherpa-4.16.0> hdrs2[2]["EXTNAME"]
'XTEND'

sherpa-4.16.0> hdrs2[2]["HDUNAME"]
'XTEND'
```

So, instead of creating `blkname` you'd have to do something like

```
found = None
for idx, hdr in hdrs.items():
    if hdr["INSTRUME"] != instrument or hdr["DETNAM"] != detnam or ...:
        continue
    found = idx
    break

if found is None:
    raise SomeSortOfError

coldata = blocks[idx]
hdrdata = hdrs[idx]
```

Reply:
I've been playing around `read_table_blocks` and the two IO backends, and I think it's a bit fragile if we're relying on the header keywords pulled from the original CalDB files... and they just don't necessarily provide sufficient information.  For example, the XMM file is absolutely bare except for the keyword I added to it when I extracted it:

```
falafel-119: dmlist xmm-ebounds-lut.fits keys
 
--------------------------------------------------------------------------------
Header keys for block EPIC-MOS_PI
--------------------------------------------------------------------------------
 
0001 EUNIT                eV                             String       energy grid units
```

and even the ACIS gratings file doesn't have a keyword to indicate the grating order...

```
falafel-122: dmlist chandra-ebounds-lut.fits keys
 
--------------------------------------------------------------------------------
Header keys for block ACIS-HEG/1
--------------------------------------------------------------------------------
 
0001 CREATOR              mkgrmf v0.2.4-1                String       Program creating this file
0002 DATE                 2024-07-15T17:08:21            String       Date and time of file creation
0003 FILTER                                              String       
0004 CHANTYPE             PI                             String       
0005 DETCHANS             8192                           Int4         
0006 HDUCLASS             OGIP                           String       
0007 HDUCLAS1             RESPONSE                       String       
0008 HDUCLAS2             EBOUNDS                        String       
0009 HDUVERS              1.3.0                          String       
0010 RMFVERSN             1992a                          String       Obsolete
0011 HDUVERS1             1.1.0                          String       Obsolete
0012 HDUVERS2             1.2.0                          String       Obsolete
0013 MISSION              AXAF                           String       Mission
0014 TELESCOP             CHANDRA                        String       Telescope
0015 INSTRUME             ACIS                           String       Instrument
0016 DETNAM               ACIS-7                         String       Detector
0017 OBS_ID               7280                           String       Observation id
0018 DS_IDENT             10.25574/07280                 String       Dataset Identifier: DOI
0019 READMODE             TIMED                          String       Read mode
0020 GRATING              HETG                           String       Grating
0021 CYCLE                P                              String       
0022 SIM_X                   -0.68282252473119           Real8        SIM focus pos
0023 SIM_Y                                   0           Real8        SIM orthogonal axis pos
0024 SIM_Z                     -187.1254020033           Real8        SIM translation stage pos
0025 CHECKSUM             A3TAD2R5A2RAA2R3               String       HDU checksum updated 2024-07-17T20:57:30
0026 DATASUM              718591150                      String       data unit checksum updated 2024-07-15T17:08:22
0027 EUNIT                keV                            String       energy grid units
```

So I'm not entirely convinced this is a more appropriate approach.  I went about handling pycrates vs. astropy separately because of the HDU name issues I was having, which you've highlighted here...

Reply:
I might be a dolt, but maybe the easiest thing is to also write the block name as a new header keyword too when for these tables?  Then I'd be able to use `read_table_blocks`, agnostic of backend.

Reply:
Yeah - the generic I/O routines are somewhat specific to Sherpa and not really designed as a general purpose library.